### PR TITLE
Add Azure firewall policy module

### PR DESCRIPTION
**Pull request by rahalan**
# Change

Adding Azure Firewall policy module and child resources

Pipeline reference
[Network: FirewallPolicies](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.firewallpolicies.yml)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: arm/Microsoft.Network/azureFirewalls/deploy.bicep [hunk 1]
```diff
@@ -215,6 +218,9 @@ resource azureFirewall 'Microsoft.Network/azureFirewalls@2021-02-01' = {
     additionalProperties: {
       'Network.DNS.EnableProxy': string(enableDnsProxy)
     }
+    firewallPolicy: {
+      id: empty(firewallPolicyIResourceId) ? null : firewallPolicyIResourceId
```
Comment (MrMCake) [inline arm/Microsoft.Network/azureFirewalls/deploy.bicep:222]:
```suggestion
      id: !empty(firewallPolicyIResourceId) ? firewallPolicyIResourceId : null
```

Just for readibility please flip

Comment (MrMCake) [reply arm/Microsoft.Network/azureFirewalls/deploy.bicep:222]: If we do it like that, would the template remove an existing firewall policy if we don't provide one with the deployment? I'm thinking that maybe it's better to have a condition for the entire object. For example:

```bicep
firewallPolicy: !empty(firewallPolicyIResourceId) ? {
      id:  firewallPolicyIResourceId
} : null
``` 

However, I guess it is up to testing whether this is actually the case

File: arm/Microsoft.Network/firewallPolicies/deploy.bicep [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  properties: {
+    basePolicy: !empty(basePolicyResourceId) ? {
+      id: basePolicyResourceId
+    } : null
+    dnsSettings: enableProxy ? {
+      enableProxy: enableProxy
+      requireProxyForNetworkRules: requireProxyForNetworkRules
+      servers: servers
+    } : null
+    insights: isEnabled ? {
```
Comment (MrMCake) [inline arm/Microsoft.Network/firewallPolicies/deploy.bicep:119]: I wonder if we should rename `isEnabled` to `insightsIsEnabled` to be more explicit. Thoughts? I know we'd deviate from the ARM reference, but then again you wouldn't know that is going on unless you check out the readme/description. @MariusStorhaug , thoughts?

File: arm/Microsoft.Network/firewallPolicies/deploy.bicep [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+      certificateAuthority: {
+        keyVaultSecretId: !empty(keyVaultSecretId) ? keyVaultSecretId : null
+        name: !empty(certificateName) ? certificateName : null
+      }
+    } : null
+  }
+}
+
+module ruleCollectionGroups_resource 'ruleCollectionGroups/deploy.bicep' = [for (ruleCollectionGroup, index) in ruleCollectionGroups: {
+  name: '${uniqueString(deployment().name, location)}-ruleCollectionGroup-${index}'
```
Comment (MrMCake) [inline arm/Microsoft.Network/firewallPolicies/deploy.bicep:157]:
```suggestion
  name: '${uniqueString(deployment().name, location)}-firewallPolicy_ruleCollectionGroup-${index}'
```

Update as per [documentation](https://github.com/Azure/ResourceModules/wiki/ModulesDesign#deployment-names).

File: arm/Microsoft.Network/firewallPolicies/deploy.bicep [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    transportSecurity: (!empty(keyVaultSecretId) || !empty(certificateName)) ? {
+      certificateAuthority: {
+        keyVaultSecretId: !empty(keyVaultSecretId) ? keyVaultSecretId : null
+        name: !empty(certificateName) ? certificateName : null
+      }
+    } : null
+  }
+}
+
+module ruleCollectionGroups_resource 'ruleCollectionGroups/deploy.bicep' = [for (ruleCollectionGroup, index) in ruleCollectionGroups: {
```
Comment (MrMCake) [inline arm/Microsoft.Network/firewallPolicies/deploy.bicep:156]:
```suggestion
module firewallPolicy_ruleCollectionGroups 'ruleCollectionGroups/deploy.bicep' = [for (ruleCollectionGroup, index) in ruleCollectionGroups: {
```
Update das per [documentation](https://github.com/Azure/ResourceModules/wiki/ModulesDesign#deployment-names)

Comment (MrMCake) [reply arm/Microsoft.Network/firewallPolicies/deploy.bicep:156]: Please also for all other resources

File: arm/Microsoft.Network/firewallPolicies/ruleCollectionGroups/deploy.bicep [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+param ruleCollections array = []
+
+@description('Optional. Customer Usage Attribution ID (GUID). This GUID must be previously registered')
+param cuaId string = ''
+
+resource firewallPolicy 'Microsoft.Network/firewallPolicies@2021-03-01' existing = {
+  name: firewallPolicyName
+}
+
+module pid_cuaId '.bicep/nested_cuaId.bicep' = if (!empty(cuaId)) {
```
Comment (MrMCake) [inline arm/Microsoft.Network/firewallPolicies/ruleCollectionGroups/deploy.bicep:20]: Lets move that up above the `existing` resource to keep that one close together with the resource that is actually referencing it

File: arm/Microsoft.Network/firewallPolicies/ruleCollectionGroups/readme.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+For remaining properties, see [FirewallPolicyRuleCollection objects](https://docs.microsoft.com/en-us/azure/templates/microsoft.network/firewallpolicies/rulecollectiongroups?tabs=json#firewallpolicyrulecollection-objects)
+
+```json
+"ruleCollections": [
+    {
+    "name": "string",
+    "priority": "int",
+    "ruleCollectionType": "string"
+    // For remaining properties, see FirewallPolicyRuleCollection objects
```
Comment (MrMCake) [inline arm/Microsoft.Network/firewallPolicies/ruleCollectionGroups/readme.md:31]: Where?

File: arm/Microsoft.Network/firewallPolicies/ruleGroups/deploy.bicep [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+param rules array = []
+
+@description('Optional. Customer Usage Attribution ID (GUID). This GUID must be previously registered')
+param cuaId string = ''
+
+resource firewallPolicy 'Microsoft.Network/firewallPolicies@2021-03-01' existing = {
+  name: firewallPolicyName
+}
+
+module pid_cuaId '.bicep/nested_cuaId.bicep' = if (!empty(cuaId)) {
```
Comment (MrMCake) [inline arm/Microsoft.Network/firewallPolicies/ruleGroups/deploy.bicep:20]: Same comment as before. Let's move that one up

File: arm/Microsoft.Network/firewallPolicies/ruleGroups/readme.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+For remaining properties, see [FirewallPolicyRule objects](https://docs.microsoft.com/en-us/azure/templates/microsoft.network/2020-04-01/firewallpolicies/rulegroups?tabs=json#firewallpolicyrule-objects)
+
+```json
+"rules": [
+    {
+    "name": "string",
+    "priority": "int",
+    "ruleType": "string"
+    // For remaining properties, see FirewallPolicyRule objects
```
Comment (MrMCake) [inline arm/Microsoft.Network/firewallPolicies/ruleGroups/readme.md:31]: where?

File: arm/Microsoft.Network/firewallPolicies/deploy.bicep [hunk 9]
```diff
@@ -0,0 +1,189 @@
+@description('Required. Name of the Firewall Policy.')
```
Comment (MrMCake) [inline arm/Microsoft.Network/firewallPolicies/deploy.bicep:1]: Please add a status badge of the pipeline running to the PR description

File: arm/Microsoft.Network/firewallPolicies/ruleGroups/readme.md (hunk 8 above)
Comment (rahalan) [reply arm/Microsoft.Network/firewallPolicies/ruleGroups/readme.md:31]: see 8 lines above

File: arm/Microsoft.Network/firewallPolicies/ruleCollectionGroups/readme.md (hunk 6 above)
Comment (rahalan) [reply arm/Microsoft.Network/firewallPolicies/ruleCollectionGroups/readme.md:31]: see 8 lines above

File: arm/Microsoft.Network/azureFirewalls/deploy.bicep (hunk 1 above)
Comment (rahalan) [reply arm/Microsoft.Network/azureFirewalls/deploy.bicep:222]: changed

File: arm/Microsoft.Network/firewallPolicies/deploy.bicep (hunk 2 above)
Comment (rahalan) [reply arm/Microsoft.Network/firewallPolicies/deploy.bicep:119]: renamed

File: arm/Microsoft.Network/firewallPolicies/deploy.bicep (hunk 3 above)
Comment (rahalan) [reply arm/Microsoft.Network/firewallPolicies/deploy.bicep:157]: done

File: arm/Microsoft.Network/firewallPolicies/deploy.bicep (hunk 4 above)
Comment (rahalan) [reply arm/Microsoft.Network/firewallPolicies/deploy.bicep:156]: done

File: arm/Microsoft.Network/firewallPolicies/ruleCollectionGroups/deploy.bicep (hunk 5 above)
Comment (rahalan) [reply arm/Microsoft.Network/firewallPolicies/ruleCollectionGroups/deploy.bicep:20]: done

File: arm/Microsoft.Network/firewallPolicies/ruleGroups/deploy.bicep (hunk 7 above)
Comment (rahalan) [reply arm/Microsoft.Network/firewallPolicies/ruleGroups/deploy.bicep:20]: done

File: arm/Microsoft.Network/firewallPolicies/deploy.bicep (hunk 9 above)
Comment (rahalan) [reply arm/Microsoft.Network/firewallPolicies/deploy.bicep:1]: [Network: FirewallPolicies](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.firewallpolicies.yml)

Comment (MrMCake) [reply arm/Microsoft.Network/firewallPolicies/deploy.bicep:1]: I meant the PR description, not this comment :D Will add it there for you

File: arm/Microsoft.Network/azureFirewalls/deploy.bicep (hunk 1 above)
Comment (MrMCake) [reply arm/Microsoft.Network/azureFirewalls/deploy.bicep]: Did you by any chance test it?

Comment (rahalan) [reply arm/Microsoft.Network/azureFirewalls/deploy.bicep]: test failed. Removed the property (reverted back to old one)

Comment (MrMCake) [reply arm/Microsoft.Network/azureFirewalls/deploy.bicep]: Alright, thanks

**Review by MrMCake — APPROVED** [review]

**Pull request merged**